### PR TITLE
restoration.denoise_bilateral and negative value image

### PR DESCRIPTION
**Pull request by odebeir**
Here is a fix for #1677 regarding  the case of negative-valued images, not fixed in  #2533
basically change the test inside .pyx and adjusted docstring

### CODE REVIEW

**Comment by jni** [comment]
@odebeir, @emmanuelle mentions that we could just offset the image values and back if there are any negative values. This seems easy enough to do? What do you think?

File: skimage/restoration/_denoise.py [hunk 1]
```diff
@@ -23,7 +23,7 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
     Parameters
     ----------
     image : ndarray
-        Input image.
+        Input image (image minimum must be >0).
```
Comment (emmanuelle) [inline skimage/restoration/_denoise.py]: It's not completely true: values must be >= 0, it's the maximum that has to be >0.

Comment (odebeir) [inline skimage/restoration/_denoise.py]: in fact is one value is <0 it breaks as well

**Comment by emmanuelle** [comment]
What I suggest is 
- substract the minimum (and add it again before returning the output, of course)
- when testing whether the maximum is zero, don't raise an error but return the (constant) input if zero
- add a test with an array of zeros and an array of negative values in the test module

**Comment by emmanuelle** [comment]
@jni would you agree?

File: skimage/restoration/_denoise.py [hunk 2]
```diff
@@ -23,7 +23,7 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
     Parameters
     ----------
     image : ndarray
-        Input image.
+        Input image (image must be >= 0 with at least one > 0 value).
```
Comment (ahojnnes) [inline skimage/restoration/_denoise.py]: How about: "Input image whose minimum value is greater than 0."

**Comment by odebeir** [comment]
I added three tests for the function:
* check if a zero array returns identity
* check if a const negative image returns identity
* check if 2 images with an offset (eventually negative pixels) give the same result (with the same offset)

the modified python wrapper _denoise pass these tests

**Comment by odebeir** [comment]
it is amaizing:
adding test actually diminish the coverage ...

File: skimage/restoration/_denoise.py [hunk 3]
```diff
@@ -56,8 +56,15 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
...
-    return _denoise_bilateral(image, win_size, sigma_range, sigma_spatial,
-                              bins, mode, cval)
+
+    offset = np.min(image)
+    offset_img = image - offset
+    if not offset_img.any():
+        # if all value identical, returns the same image
+        return image
+    else:
+        return _denoise_bilateral(offset_img+1, win_size, sigma_range, sigma_spatial,
```
Comment (emmanuelle) [inline skimage/restoration/_denoise.py]: why the +1 and -1? I don't understand.

File: skimage/restoration/_denoise_cy.pyx [hunk 4]
```diff
@@ -59,7 +59,7 @@ def _denoise_bilateral(image, Py_ssize_t win_size, sigma_range,
     # if image.max() is 0, then dist_scale can have an unverified value
     # and color_lut[<int>(dist * dist_scale)] may cause a segmentation fault
     # so we verify we have a positive image and that the max is not 0.0.
-    if image.min() < 0.0:
+    if image.min() <= 0.0:
```
Comment (emmanuelle) [inline skimage/restoration/_denoise_cy.pyx]: I think that now that we substract the min value, as you did in the Python function, it's ok to have zero values.

File: skimage/restoration/tests/test_denoise.py [hunk 5]
```diff
@@ -110,6 +110,29 @@ def test_denoise_tv_bregman_3d():
     assert img[30:45, 5:15].std() > out1[30:45, 5:15].std()
     assert out1[30:45, 5:15].std() > out2[30:45, 5:15].std()
 
+def test_denoise_bilateral_null():
+    img = np.zeros((50, 50))
+    out = restoration.denoise_bilateral(img)
+
+    # image full of zeros should return identity
+    assert_equal(out, img)
+
```
Comment (emmanuelle) [inline skimage/restoration/tests/test_denoise.py]: please add a blank line (there should be two blank lines between functions)

File: skimage/restoration/tests/test_denoise.py [hunk 6]
```diff
@@ -110,6 +110,29 @@ def test_denoise_tv_bregman_3d():
...
+def test_denoise_bilateral_negative():
+    img = -np.ones((50, 50))
+    out = restoration.denoise_bilateral(img)
+
+    # image with only negative values should be ok
+    assert_equal(out, img)
+
```
Comment (emmanuelle) [inline skimage/restoration/tests/test_denoise.py]: and here as well

File: skimage/restoration/tests/test_denoise.py [hunk 7]
```diff
@@ -110,6 +110,29 @@ def test_denoise_tv_bregman_3d():
...
+def test_denoise_bilateral_negative2():
+    img = np.ones((50, 50))
+    img[2,2] = 2
```
Comment (emmanuelle) [inline skimage/restoration/tests/test_denoise.py]: some PEP8 here: ``img[2, 2]``

File: skimage/restoration/tests/test_denoise.py [hunk 8]
```diff
@@ -110,6 +110,29 @@ def test_denoise_tv_bregman_3d():
...
+
+    out1 = restoration.denoise_bilateral(img)
+    out2 = restoration.denoise_bilateral(img-10) #contains negative values
```
Comment (emmanuelle) [inline skimage/restoration/tests/test_denoise.py]: pep8 ``img - 10``

File: skimage/restoration/tests/test_denoise.py [hunk 9]
```diff
@@ -110,6 +110,29 @@ def test_denoise_tv_bregman_3d():
...
+
+    # 2 images with a given offset should give the same result (with the same offset)
+    assert_equal(out1, out2+10)
```
Comment (emmanuelle) [inline skimage/restoration/tests/test_denoise.py]: ``out2 + 10``

**Comment by emmanuelle** [comment]
yeah, coveralls is doing strange things these days. Nevermind, your new tests are indeed useful. I just added a couple of comments.

File: skimage/restoration/_denoise.py (hunk 3 above)
Comment (odebeir) [inline skimage/restoration/_denoise.py]: well, it was related to a previous correction (commit 8d6ba04),
now, with the offset in the python wrapper it is indeed not needed anymore, so I reverted the .pyx to its initial state and removed the +1/-1

File: skimage/restoration/tests/test_denoise.py (hunk 9 above)
Comment (odebeir) [inline skimage/restoration/tests/test_denoise.py]: rem: I autopep8 the complete test_denoise.py file (other small changes made)

File: skimage/restoration/_denoise.py [hunk 10]
```diff
@@ -56,8 +56,15 @@ def denoise_bilateral(image, win_size=5, sigma_range=None, sigma_spatial=1,
...
-    return _denoise_bilateral(image, win_size, sigma_range, sigma_spatial,
-                              bins, mode, cval)
+
+    offset = np.min(image)
+    offset_img = image - offset
+    if not offset_img.any():
+        # if all value identical, returns the same image
+        return image
+    else:
+        return _denoise_bilateral(offset_img, win_size, sigma_range, sigma_spatial,
```
Comment (ahojnnes) [inline skimage/restoration/_denoise.py]: This is changing the existing behavior and I don't see a reason why to subtract the min from the image. If you want to test for uniqueness, do a `image.min() - image.max() == 0` test.

Comment (odebeir) [inline skimage/restoration/_denoise.py]: basically the idea was to enable the process on image having <0 values...
using the offset ensures it as suggested by @emmanuelle

Comment (ahojnnes) [inline skimage/restoration/_denoise.py]: I agree that negative values should be supported, but we shouldn't mess with the range of values in the image. This should be easily possible by changing the Cython implementation accordingly.

Comment (odebeir) [inline skimage/restoration/_denoise.py]: I guess so, the solution was only a (quick) fix.

**Comment by stefanv** [comment]
@ahojnnes @odebeir What is the next step forward on this one?

**Comment by jni** [comment]
Oops, this one fell through the cracks... @odebeir do you think you could rebase on the current master? Then we can, hopefully, do a quick final review and get it merged!

**Comment by odebeir** [comment]
@jni, I am a bit confused here, I am not sure to be able to do the suggested maneuver ...

**Comment by soupault** [comment]
Fixed in #2533.

**Comment by jni** [comment]
@soupault No, that fixed the case of single-value images. This fixes the case of negative-valued images.

**Comment by sciunto** [comment]
Already fixed by  #2533

**Comment by sciunto** [comment]
I updated the first message and the title to avoid confusion

**Comment by grlee77** [comment]
This PR seems like a useful, straightforward addition that never quite got merged. I fixed the merge conflicts and have revived it in #5527
